src/names/applications.cpp

```diff
@@ -0,0 +1,190 @@
+            std::regex domainPattern("^(xn--)?[a-z0-9]+(-[a-z0-9]+)*$");
+            if (!std::regex_match (label, domainPattern))
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            // Reject digits-only labels
+            std::regex digitsOnly("^[0-9]+$");
+            if (std::regex_match (label, digitsOnly))
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            return NameNamespace::Domain;
+        }
+        case NameNamespace::DomainData:
+        {
+            const std::string label = name.substr(purportedLen);
+
+            if (!IsStringValid (label, NameEncoding::ASCII))
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            if (label.length() < 1)
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            return NameNamespace::DomainData;
+        }
+        case NameNamespace::Identity:
+        {
+            const std::string label = name.substr(purportedLen);
+
+            if (!IsStringValid (label, NameEncoding::ASCII))
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            if (label.length() < 1)
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            // Max id/ identifier length is 255 chars according to wiki spec.
+            // But we don't need to check for this, because that's also the max
+            // length of an identifier under the Namecoin consensus rules.
+
+            // Same as d/ regex but without IDN prefix.
+            // TODO: this doesn't exactly match the https://wiki.namecoin.org spec.
+            std::regex identityPattern("^[a-z0-9]+(-[a-z0-9]+)*$");
+            if (!std::regex_match (label, identityPattern))
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            return NameNamespace::Identity;
+        }
+        case NameNamespace::IdentityData:
+        {
+            const std::string label = name.substr(purportedLen);
+
+            if (!IsStringValid (label, NameEncoding::ASCII))
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            if (label.length() < 1)
+            {
+                return NameNamespace::NonStandard;
+            }
+
+            return NameNamespace::IdentityData;
+        }
+        case NameNamespace::NonStandard:
+            return NameNamespace::NonStandard;
+    }
+
+    /* NameNamespace values are only ever created internally in the binary
+    and not received externally (except as string).  Thus it should never
+    happen (and if it does, is a severe bug) that we see an unexpected
+    value.  */
+    assert (false);
+}
+
+NameNamespace
+NamespaceFromName (const valtype& data)
+{
+    std::string name = std::string (data.begin (), data.end ());
```

Comment:
Can you use `DecodeName` here instead?  Even though it will not reduce code volume, it seems like the "proper" function to convert between `valtype` and `std::string`.

Reply:
Agree in principle.  I didn't do this because `DecodeName` requires a `NameEncoding`, and at this point in the code we don't yet know that the name is valid ASCII or UTF-8.  Would it be okay to add a new `NameEncoding::BINARY` to the `enum`, which acts as a pass-through with no validation?  This would simply a lot of my other name GUI code too.

Reply:
Why not just use UTF-8, which seems like a reasonable encoding in any case for a valid name?

Reply:
In this particular case that might not hurt anything, but there are a bunch of other places in the name GUI code where I convert between `valtype` and `std::string` where binary data very well may show up (e.g. values of `d/` names, which may in the future start using CBOR encoding).  If we are agreed that "manual" conversion between those two types is undesirable for readability, I'd like to fix it everywhere, not just in the places where we are confident that the data is valid UTF-8.

Reply:
I see.  Well at the moment, basically `valtype` and `std::string` are not supposed to be directly convertible, but one is encoded while the other is not (similar to `str` and `bytes` in Python).  So the conversion between them is using a particular encoding.

`std::string` is used when interfacing with the user, e.g. through RPC of the Qt UI.  Are those places supposed to handle binary directly in the future?

Reply:
I'm not sure I understand; `std::string` is not (AFAIK) aware of encoding; it's just a sequence of single-byte characters.  In contrast, Python `str` is aware of encoding.  You can put invalid UTF-8 into a `std::string` and nothing bad will happen.  Is what you're describing just a non-standard convention used by Bitcoin Core to try to improve type-safety (in which case I guess I'm fine with abiding by it; just never ran into it before)?

Reply:
This is just a convention I have in mind for the use of `valtype` vs `std::string`, based on how it evolved and how I think it makes sense for internal use especially with names/values.  This can of course be changed if you have arguments for why you want to use `std::string` for binary values (but then the main question probably is - why not use `valtype` instead, which is basically what is meant for binary).

Reply:
This sounds like a type-safety convention by another name, which I'm fine with.  I will refactor this PR to avoid such conversions.

Reply:
Fixed.